### PR TITLE
feat: display ERC20 and ERC721 token details returns by decoding api

### PR DESCRIPTION
**Pull request by jpuri**
## **Description**

Add estimated changes display for ERC20 token and NFTs.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/3614

## **Manual testing steps**

1. Enable signature decoding api
2. On test dapp submit permit request
3. Check simulation section

## **Screenshots/Recordings**
Screenshot 2024-11-08 at 12 19 40 AM
Screenshot 2024-11-08 at 12 18 24 AM
Screenshot 2024-11-08 at 12 16 13 AM

## **Pre-merge author checklist**

- [X] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by pedronfigueiredo — DISMISSED** [review]

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx [hunk 1]
```diff
@@ -1,7 +1,95 @@
...
+};
+
+const DecodedSimulation: React.FC<object> = () => {
+  const t = useI18nContext();
+  const { currentConfirmation } = useConfirmContext<SignatureRequestType>();
+  const chainId = currentConfirmation.chainId as Hex;
+  const { decodingLoading, decodingData } = currentConfirmation;
+
+  let stateChangeFragment: ReactNode[] = [];
+  if (decodingData?.stateChanges) {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx:77]: Minor, is there a need for this `if` and `let` as we could just `map` from the `stateChanges` with a `[]` fallback?

```
const stateChanges = decodingData?.stateChanges ?? [];
const stateChangeFragment = stateChanges.map(...);`
```

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx [hunk 2]
```diff
@@ -1,7 +1,95 @@
...
+  TextColor,
+  TextVariant,
+} from '../../../../../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../../../../../hooks/useI18nContext';
+import { SignatureRequestType } from '../../../../../../types/confirm';
+import { useConfirmContext } from '../../../../../../context/confirm';
+import StaticSimulation from '../../../shared/static-simulation/static-simulation';
+import PermitSimulationValueDisplay from '../value-display/value-display';
+
+const getStateChangeLabelMap = (
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx:23]: Minor, if this is static, is it simpler to define in a constant using just the translation keys, then call `t` once on use`?

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/value-display/value-display.tsx [hunk 3]
```diff
@@ -50,11 +51,25 @@ type PermitSimulationValueDisplayParams = {
 
   /** The tokenId for NFT */
   tokenId?: string;
+
+  /** True if value is being credited to wallet */
+  credit?: boolean;
+
+  /** True if value is being debited to wallet */
+  debit?: boolean;
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/value-display/value-display.tsx:59]: If there is only one value, won't this two properties always alternate, in which case could we have a single `isDebit`?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx (hunk 2 above)
Comment (jpuri) [reply ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx:23]: Calling `t` for strings is useful to find when strings are not used in translation.

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx (hunk 1 above)
Comment (jpuri) [reply ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx:77]: fixed

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/value-display/value-display.tsx (hunk 3 above)
Comment (jpuri) [reply ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/value-display/value-display.tsx:59]: A value may be none of debit and credit, thus is it more clear to have 2 fields. These are not complete mutually exclusive.

**Review by pedronfigueiredo — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

File: ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/value-display/value-display.tsx (hunk 3 above)
Comment (matthewwalsh0) [reply ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/value-display/value-display.tsx:59]: Could the third state be handled using the `undefined` value of the field? Is that more idiomatic to JS?

**Pull request merged**